src/container.cpp

```diff
@@ -291,50 +289,50 @@ ReturnValue Container::queryAdd(int32_t index, const Thing& thing, uint32_t coun
 	}
 
 	if (!hasBitSet(FLAG_NOLIMIT, flags)) {
-		while (cylinder) {
-			if (cylinder == &thing) {
+		while (parent) {
+			if (parent == thing) {
 				return RETURNVALUE_THISISIMPOSSIBLE;
 			}
 
-			if (dynamic_cast<const Inbox*>(cylinder)) {
+			if (std::dynamic_pointer_cast<const Inbox>(parent)) {
 				return RETURNVALUE_CONTAINERNOTENOUGHROOM;
 			}
 
-			cylinder = cylinder->getParent();
+			parent = parent->getParent();
 		}
 
 		if (index == INDEX_WHEREEVER && size() >= capacity() && !hasPagination()) {
 			return RETURNVALUE_CONTAINERNOTENOUGHROOM;
 		}
 	} else {
-		while (cylinder) {
-			if (cylinder == &thing) {
+		while (parent) {
+			if (parent == thing) {
 				return RETURNVALUE_THISISIMPOSSIBLE;
 			}
 
-			cylinder = cylinder->getParent();
+			parent = parent->getParent();
 		}
 	}
 
-	const Cylinder* const topParent = getTopParent();
+	auto const topParent = getTopParent();
 	if (actor && getBoolean(ConfigManager::ONLY_INVITED_CAN_MOVE_HOUSE_ITEMS)) {
-		if (const HouseTile* const houseTile = dynamic_cast<const HouseTile*>(topParent->getTile())) {
+		if (auto houseTile = std::dynamic_pointer_cast<const HouseTile>(topParent->getTile())) {
 			if (!topParent->getCreature() && !houseTile->getHouse()->isInvited(actor->getPlayer())) {
 				return RETURNVALUE_PLAYERISNOTINVITED;
 			}
 		}
 	}
 
-	if (topParent != this) {
-		return topParent->queryAdd(INDEX_WHEREEVER, *item, count, flags | FLAG_CHILDISOWNER, actor);
+	if (topParent.get() != this) {
+		return topParent->queryAdd(INDEX_WHEREEVER, item, count, flags | FLAG_CHILDISOWNER, actor);
 	}
 	return RETURNVALUE_NOERROR;
 }
 
-ReturnValue Container::queryMaxCount(int32_t index, const Thing& thing, uint32_t count, uint32_t& maxQueryCount,
-                                     uint32_t flags) const
+ReturnValue Container::queryMaxCount(int32_t index, std::shared_ptr<const Thing> thing, uint32_t count,
+                                     uint32_t& maxQueryCount, uint32_t flags) const
 {
-	const Item* item = thing.getItem();
+	auto item = thing->getItem();
```

Comment:
Could this be `const auto`
```suggestion
	const auto item = thing->getItem();
```